### ISSUE TITLE
Repacking not running parallelized

### ISSUE THREAD

**Issue by oberstet** [reorg/pg_repack#102]
I am trying to repack a data-warehouse (PostgreSQL 9.5.4) with approx. 50.000 tables (due to partitioning) and 20 TB of data.

The machine running the DB has massive CPU (48 cores with 3TB RAM) and IOPS (500k/sec random 8kB).

So I started:

```
pg_repack --jobs=16 -d adr
```

to repack using 16 connections in parallel. Indeed, I can see these connections being opened, but all but one connection sit idle all the time.

Since there are 50k tables (well, partitions with indices), there should be enough room for parallelization.

What am I doing wrong?

**Comment by oberstet** [comment]
Only 1 process working:

```
XXXX (debian jessie/sid 64bit / Linux 3.19.0-78-generic)                                                                                                Uptime: 2:09:05

CPU  [||                                         4.6%]   CPU       4.6%  nice:     0.0%  ctx_sw: 69991   MEM     26.3%  active:     190G   SWAP      0.0%   LOAD    48-core
MEM  [|||||||||||                               26.3%]   user:     2.7%  irq:      0.0%  inter:  70771   total:  2.95T  inactive:  20.6G   total:   74.4G   1 min:    1.61
SWAP [                                             0%]   system:   1.9%  iowait:   0.0%  sw_int:  1660   used:    795G  buffers:   90.3M   used:        0   5 min:    1.52
                                                         idle:    95.4%  steal:    0.0%                  free:   2.18T  cached:     217G   free:    74.4G   15 min:   1.42

NETWORK     Rx/s   Tx/s   TASKS 612 (619 thr), 4 run, 608 slp, 0 oth sorted automatically by cpu_percent, flat view
eth0        47Kb   10Kb
eth1       382Kb 55.5Mb     CPU%  MEM%  VIRT   RES   PID USER        NI S     TIME+   R/s   W/s Command 
lo          25Kb   25Kb     99.4   0.0  762G 49.5M 44483 postgres     0 D  45:03.31     0     0 adr: postgres adr [local] CREATE TABLE AS
p11p1         0b     0b      9.5   0.0  425M 28.9M 13776 oberstet     0 R   0:28.80     0     0 /opt/cpy2712/bin/python /opt/cpy2712/bin/glances
                             8.6   0.0  762G 24.9M 30837 postgres     0 R   1:19.45     0     0 adr: checkpointer process   
DISK I/O     R/s    W/s      2.1   0.0  762G 4.32M 39241 postgres     0 S   0:40.30     0     0 adr: wal writer process   
md0            0      0      1.8   0.0 43.5M 2.79M 44918 postgres     0 S   0:28.73     0     0 postgres@10.1.1.14:22 [mux]
md1            0    20K      0.3   0.0 34.3M 3.77M 39243 postgres     0 S  28:12.59     0     0 adr: archiver process   archiving 000000040000C0AE0000007D
md127      79.5M   459M      0.3   0.0     0     0 46781 root         0 S   0:01.17     0     0 kworker/18:0
md3            0      0      0.3   0.0  762G 6.62M 30184 postgres     0 S  20:21.77     0     0 /opt/pg954/bin/postgres -D /data/adr
nvme0n1    4.97M  28.7M      0.3   0.0     0     0     8 root         0 S   0:13.40     0     0 rcu_sched
nvme10n1   4.97M  28.7M      0.3   0.0     0     0 12251 root         0 S   0:00.50     0     0 kworker/19:2
nvme11n1   4.96M  28.7M      0.3   0.0     0     0   419 root         0 S   0:00.63     0     0 kworker/30:1
nvme12n1   4.97M  28.7M      0.3   0.0     0     0   181 root         0 S   0:01.39     0     0 rcuos/24
...
```

**Comment by oberstet** [comment]
The log seems to suggest that it runs sequentially:

```
INFO: repacking table "x_adr_core_partitions.t_rating_ereignis_201204"
LOG: Initial worker 0 to build index: CREATE INDEX index_399341200 ON repack.table_191339463 USING btree (q_nrkunde)
LOG: Initial worker 1 to build index: CREATE UNIQUE INDEX index_191339469 ON repack.table_191339463 USING btree (rating_key, event_tsp, version_from)
LOG: Initial worker 2 to build index: CREATE INDEX index_273061757 ON repack.table_191339463 USING btree (segment_key) WHERE (segment_key <> ALL (ARRAY[16, 18]))
LOG: Command finished in worker 2: CREATE INDEX index_273061757 ON repack.table_191339463 USING btree (segment_key) WHERE (segment_key <> ALL (ARRAY[16, 18]))
LOG: Command finished in worker 1: CREATE UNIQUE INDEX index_191339469 ON repack.table_191339463 USING btree (rating_key, event_tsp, version_from)
LOG: Command finished in worker 0: CREATE INDEX index_399341200 ON repack.table_191339463 USING btree (q_nrkunde)
```
and

```
INFO: repacking table "x_adr_core_partitions.t_rating_ereignis_201205"
LOG: Initial worker 0 to build index: CREATE INDEX index_399341199 ON repack.table_191339472 USING btree (q_nrkunde)
LOG: Initial worker 1 to build index: CREATE UNIQUE INDEX index_191339478 ON repack.table_191339472 USING btree (rating_key, event_tsp, version_from)
LOG: Initial worker 2 to build index: CREATE INDEX index_273061759 ON repack.table_191339472 USING btree (segment_key) TABLESPACE ext1 WHERE (segment_key <> ALL (ARRAY[16, 18]))
LOG: Command finished in worker 2: CREATE INDEX index_273061759 ON repack.table_191339472 USING btree (segment_key) TABLESPACE ext1 WHERE (segment_key <> ALL (ARRAY[16, 18]))
LOG: Command finished in worker 1: CREATE UNIQUE INDEX index_191339478 ON repack.table_191339472 USING btree (rating_key, event_tsp, version_from)
LOG: Command finished in worker 0: CREATE INDEX index_399341199 ON repack.table_191339472 USING btree (q_nrkunde)
```

and so on

**Comment by MasahikoSawada** [comment]
As pg_repack document says, only rebuilding indexes on a table phase is executed in parallel.

```
-j, --jobs
 Create the specified number of extra connections to PostgreSQL, and use these extra connections to parallelize the rebuild of indexes on each table.
Parallel index builds are only supported for full-table repacks, not with --index or --only-indexes options. 
If your PostgreSQL server has extra cores and disk I/O available, this can be a useful way to speed up pg_repack.
```

**Comment by dvarrazzo** [comment]
It seems no issue then, right?

**Comment by oberstet** [comment]
Well, I've written my own thing to parallelize rebuilding over table/index partitions.

**Comment by pdomagala** [comment]
@oberstet and how it worked? any issues?

**Comment by oberstet** [comment]
the "thing" (named "SQLBalancer") I wrote for that company was built using PL/pgSQL and a Python  daemon and able to parallelize SQL based jobs by spreading over many DB connections . the DWH was running on a 96 cores / 3 TB RAM / 32TB NVMe machine in the end, and using ^ approach, we were able to saturate the HW capabilities also for table/index reorgs. millions of IOPS driven from SQL. so "no problems".
